src/H5Fint.c

```diff
@@ -1624,8 +1624,7 @@ H5F__dest(H5F_t *f, hbool_t flush)
         HDONE_ERROR(H5E_FILE, H5E_CANTINIT, FAIL, "problems closing file");
     f->shared = NULL;
 
-    if (ret_value >= 0)
-        f = H5FL_FREE(H5F_t, f);
+    f = H5FL_FREE(H5F_t, f);
```

Comment:
Always free the H5F_t structure even if H5F__dset fails. Otherwise, this just leads to a memory leak.

Reply:
Revised this based on [this comment](https://github.com/HDFGroup/hdf5/blob/develop/hl/test/test_file_image.c#L368-L380). Always freeing the file structure was leading to some double frees when a file fails to be closed from H5Fclose and then the library later goes to free the structure to clear out IDs. Instead, only free the structure on failure if the call to `H5F__dest` came from `H5F_open`. Otherwise, it'll be freed later on.